Mathlib/CategoryTheory/Subobject/ArtinianObject.lean

```diff
@@ -0,0 +1,162 @@
+term in `ObjectProperty C` which allows to study the stability
+properties of artinian objects. For statements regarding
+specific objects, it is advisable to use the type class
+`IsArtinianObject` instead. -/
+@[stacks 0FCF]
+def isArtinianObject : ObjectProperty C :=
+  fun X ↦ WellFoundedLT (Subobject X)
+
+variable (X Y : C)
+
+/-- An object `X` in a category `C` is Artinian if `Subobject X`
+satisfies the ascending chain condition. -/
+@[stacks 0FCF]
+abbrev IsArtinianObject : Prop := isArtinianObject.Is X
+
+instance [IsArtinianObject X] : WellFoundedLT (Subobject X) :=
+  isArtinianObject.prop_of_is X
+
+lemma isArtinianObject_iff_antitone_chain_condition :
+    IsArtinianObject X ↔ ∀ (f : ℕ →o (Subobject X)ᵒᵈ),
+      ∃ (n : ℕ), ∀ (m : ℕ), n ≤ m → f n = f m := by
+  dsimp only [IsArtinianObject]
+  rw [ObjectProperty.is_iff, isArtinianObject,
+    ← wellFoundedGT_dual_iff,
+    wellFoundedGT_iff_monotone_chain_condition]
+
+variable {X} in
+lemma antitone_chain_condition_of_isArtinianObject
+    [IsArtinianObject X] (f : ℕ →o (Subobject X)ᵒᵈ) :
+    ∃ (n : ℕ), ∀ (m : ℕ), n ≤ m → f n = f m :=
+  (isArtinianObject_iff_antitone_chain_condition X).1 inferInstance f
+
+lemma isArtinianObject_iff_not_strictAnti :
+    IsArtinianObject X ↔ ∀ (f : ℕ → Subobject X), ¬ StrictAnti f := by
+  refine ⟨fun _ ↦ not_strictAnti_of_wellFoundedLT, fun h ↦ ?_⟩
+  dsimp only [IsArtinianObject]
+  rw [ObjectProperty.is_iff, isArtinianObject, WellFoundedLT,
+    isWellFounded_iff, RelEmbedding.wellFounded_iff_no_descending_seq]
+  exact ⟨fun f ↦ h f.toFun (fun a b h ↦ f.map_rel_iff.2 h)⟩
+
+variable {X} in
+lemma not_strictAnti_of_isArtinianObject
+    [IsArtinianObject X] (f : ℕ → Subobject X) :
+    ¬ StrictAnti f :=
+  (isArtinianObject_iff_not_strictAnti X).1 inferInstance f
+
+lemma isArtinianObject_iff_isEventuallyConstant :
+    IsArtinianObject X ↔ ∀ (F : ℕ ⥤ (MonoOver X)ᵒᵖ),
+      IsFiltered.IsEventuallyConstant F := by
+  rw [isArtinianObject_iff_antitone_chain_condition]
+  constructor
+  · intro h G
+    obtain ⟨n, hn⟩ := h ⟨_, (G ⋙ (Subobject.equivMonoOver X).inverse.op ⋙
+      (orderDualEquivalence _).inverse).monotone⟩
+    refine ⟨n, fun m hm ↦ ?_⟩
+    rw [← isIso_unop_iff, MonoOver.isIso_iff_subobjectMk_eq]
+    exact (hn m (leOfHom hm)).symm
+  · intro h F
+    obtain ⟨n, hn⟩ := h (F.monotone.functor ⋙ (orderDualEquivalence _).functor ⋙
+      Subobject.representative.op)
+    refine ⟨n, fun m hm ↦ Eq.symm ?_⟩
+    simpa [isIso_op_iff, isIso_iff_of_reflects_iso, PartialOrder.isIso_iff_eq]
+      using hn (homOfLE hm)
+
+variable {X} in
+lemma isEventuallyConstant_of_isArtinianObject [IsArtinianObject X]
+    (F : ℕ ⥤ (MonoOver X)ᵒᵖ) : IsFiltered.IsEventuallyConstant F :=
+  (isArtinianObject_iff_isEventuallyConstant X).1 inferInstance F
+
+variable {X Y}
+
+lemma isArtinianObject_of_isZero (hX : IsZero X) : IsArtinianObject X := by
+  rw [isArtinianObject_iff_antitone_chain_condition]
+  have := Subobject.subsingleton_of_isZero hX
```

Comment:
Probably not work for this PR, but is there a reason why  `Subobject.subsingleton_of_isZero` is not an instance? Perfs?

Reply:
Right, I keep forgetting that the `Is<Co/Limits>` are not classes. Thanks for the reminder!